File: sources/CETech/Log.cs
using System;

namespace CETech
{
    /// <summary>
    /// Log system.
    /// </summary>
    public static class Log
    {
        /// <summary>
        /// Log level enum.
        /// </summary>
        public enum Level
        {
            /// <summary>
            /// Informatin message
            /// </summary>
            Info,

            /// <summary>
            /// Warning message
            /// </summary>
            Warning,
            
            /// <summary>
            /// Error message
            /// </summary>
            Error,


            /// <summary>
            /// Debug message
            /// </summary>
            Debug
        }

        /// <summary>
        /// Log handler
        /// </summary>
        /// <param name="level">Log level</param>
        /// <param name="time">Time</param>
        /// <param name="workerId">Worker id</param>
        /// <param name="where">Where</param>
        /// <param name="msg">Message</param>
        /// <param name="args">Message format arguments</param>
        public delegate void LogHandler(Level level, DateTime time, int workerId, string where, string msg, params object[] args);
        
        /// <summary>
        /// Main log event
        /// </summary>
        public static event LogHandler LogEvent;

        /// <summary>
        /// Log info message
        /// </summary>
        /// <param name="where">Where</param>
        /// <param name="msg">Message</param>
        /// <param name="args">Message format arguments</param>
        public static void Info(string where, string msg, params object[] args)
        {
            OnLog(Level.Info, DateTime.Now, 0, where, msg, args);
        }


        /// <summary>
        /// Warning info message
        /// </summary>
        /// <param name="where">Where</param>
        /// <param name="msg">Message</param>
        /// <param name="args">Message format arguments</param>
        public static void Warning(string where, string msg, params object[] args)
        {
            OnLog(Level.Warning, DateTime.Now, 0, where, msg, args);
        }

        /// <summary>
        /// Error info message
        /// </summary>
        /// <param name="where">Where</param>
        /// <param name="msg">Message</param>
        /// <param name="args">Message format arguments</param>
        public static void Error(string where, string msg, params object[] args)
        {
            OnLog(Level.Error, DateTime.Now, 0, where, msg, args);
        }

        /// <summary>
        /// Debug info message
        /// </summary>
        /// <param name="where">Where</param>
        /// <param name="msg">Message</param>
        /// <param name="args">Message format arguments</param>
        public static void Debug(string where, string msg, params object[] args)
        {
#if DEBUG
            OnLog(Level.Debug, DateTime.Now, 0, where, msg, args);
#endif
        }

        /// <summary>
        /// On log
        /// </summary>
        /// <param name="level">Log level</param>
        /// <param name="time">Time</param>
        /// <param name="workerId">Worker id</param>
        /// <param name="where">Where</param>
        /// <param name="msg">Message</param>
        /// <param name="args">Message format arguments</param>
        private static void OnLog(Level level, DateTime time, int workerId, string @where, string msg, params object[] args)
        {
            var handler = LogEvent;
            if (handler != null) handler(level, time, workerId, @where, msg, args);
        }

    }
}

File: src/corelib/private/log.c
#include <stdio.h>
#include <stdint.h>
#include <corelib/log.h>
#include <corelib/module.h>
#include <corelib/api_system.h>
#include <corelib/task.h>

#include "corelib/macros.h"


#define LOG_WHERE "log_system"

enum {
    MAX_HANDLERS = 32
};

static struct global {
    ct_log_handler_t handlers[MAX_HANDLERS];
    void *handlers_data[MAX_HANDLERS];

    char handlers_count;
} _G = {{0}};


static void vlog(const enum ct_log_level level,
                 const char *where,
                 const char *format,
                 va_list va) {

    //CE_ASSERT("logsystem", _globals.data != nullptr);

    char msg[4096];     //!< Final msg.
    vsnprintf(msg, 4096, format, va);

    time_t tm = time(NULL);

    for (uint8_t i = 0; i < _G.handlers_count; ++i) {
        _G.handlers[i](level, tm, ct_task_a0->worker_id(),
                       where, msg, _G.handlers_data[i]);
    }
}

static void log_info_va(const char *where,
                        const char *format,
                        va_list va) {
    vlog(LOG_INFO, where, format, va);
}


static void log_info(const char *where,
                     const char *format,
                     ...) {
    va_list args;

    va_start(args, format);
    log_info_va(where, format, args);
    va_end(args);
}

static void log_warning_va(const char *where,
                           const char *format,
                           va_list va) {
    vlog(LOG_WARNING, where, format, va);
}

static void log_warning(const char *where,
                        const char *format,
                        ...) {
    va_list args;

    va_start(args, format);
    log_warning_va(where, format, args);
    va_end(args);
}

static void log_error_va(const char *where,
                         const char *format,
                         va_list va) {
    vlog(LOG_ERROR, where, format, va);
}

static void log_error(const char *where,
                      const char *format,
                      ...) {
    va_list args;

    va_start(args, format);
    log_error_va(where, format, args);
    va_end(args);
}

static void log_debug_va(const char *where,
                         const char *format,
                         va_list va) {

    vlog(LOG_DBG, where, format, va);
}

static void log_debug(const char *where,
                      const char *format,
                      ...) {
    va_list args;

    va_start(args, format);
    log_debug_va(where, format, args);
    va_end(args);
}


static void log_register_handler(ct_log_handler_t handler,
                                 void *data) {
    const uint8_t idx = _G.handlers_count++;

    _G.handlers[idx] = handler;
    _G.handlers_data[idx] = data;
}



void ct_log_stdout_yaml_handler(enum ct_log_level level,
                                time_t time,
                                char worker_id,
                                const char *where,
                                const char *msg,
                                void *data);

void ct_log_stdout_handler(enum ct_log_level level,
                                time_t time,
                                char worker_id,
                                const char *where,
                                const char *msg,
                                void *data);


static struct ct_log_a0 log_a0 = {
        .stdout_handler = &ct_log_stdout_handler,
        .stdout_yaml_handler = &ct_log_stdout_yaml_handler,
        .register_handler = log_register_handler,
        .info_va = log_info_va,
        .info = log_info,
        .warning_va = log_warning_va,
        .warning = log_warning,
        .error_va = log_error_va,
        .error = log_error,
        .debug_va = log_debug_va,
        .debug = log_debug
};

struct ct_log_a0 *ct_log_a0 = &log_a0;

CETECH_MODULE_DEF(
        log,
        {
            CT_UNUSED(api);
        },
        {
            CT_UNUSED(reload);

        },
        {
            CT_UNUSED(reload);
            CT_UNUSED(api);
        }
)

